python/ray/scripts/symmetric_run.py

```diff
@@ -0,0 +1,194 @@
+def symmetric_run(address: str, nnodes: int, execute_on_head: List[str]):
+    """Start a Ray cluster and run a script only on the head node.
+
+    This command should be executed symmetrically on all nodes in the cluster.
+
+    Arguments:
+        address: The address of the Ray cluster.
+        nnodes: The number of nodes in the cluster.
+        execute_on_head: The command to run on the head node.
+    """
+
+    # 1. Parse address and check if we are on the head node.
+    try:
+        parsed_address = urlparse(f"//{address}")
+        head_ip = parsed_address.hostname
+        head_port = parsed_address.port
+    except Exception:
+        raise click.ClickException(f"Invalid address format: {address}")
+
+    if not head_ip or not head_port:
+        raise click.ClickException(
+            f"Invalid address: {address}. Must be in format <host>:<port>"
+        )
+
+    try:
+        # AF_UNSPEC allows resolving both IPv4 and IPv6
+        addrinfo = socket.getaddrinfo(
+            head_ip, head_port, socket.AF_UNSPEC, socket.SOCK_STREAM
+        )
+        resolved_head_ip = addrinfo[0][4][0]
+    except socket.gaierror:
+        raise click.ClickException(f"Could not resolve hostname: {head_ip}")
+
+    my_ips = []
+    for iface, addrs in psutil.net_if_addrs().items():
+        for addr in addrs:
+            # Look for AF_INET (IPv4) or AF_INET6 (IPv6)
+            if addr.family in [
+                socket.AddressFamily.AF_INET,
+                socket.AddressFamily.AF_INET6,
+            ]:
+                my_ips.append(addr.address)
+
+    # Add localhost ips if we are running on a single node.
+    if nnodes == 1:
+        my_ips.extend(["127.0.0.1", "::1"])
+
+    is_head = resolved_head_ip in my_ips
+
+    result = None
+    # 2. Start Ray and run commands.
+    try:
+        if is_head:
+            # On the head node, start Ray, run the command, then stop Ray.
+            click.echo("On head node. Starting Ray cluster head...")
+            # Start Ray head. This runs in the background and hides output.
+            subprocess.run(
+                [
+                    "ray",
+                    "start",
+                    "--head",
+                    f"--node-ip-address={resolved_head_ip}",
+                    f"--port={head_port}",
+                ],
+                check=True,
+                capture_output=True,
+            )
+            click.echo("Head node started.")
+            click.echo("=======================")
+            if nnodes > 1 and not check_cluster_ready(nnodes):
+                raise click.ClickException(
+                    "Timed out waiting for other nodes to start."
+                )
+
+            # Run the user command if provided.
+            if execute_on_head:
+                click.echo(f"Running command on head node: {' '.join(execute_on_head)}")
+                click.echo("=======================")
+                result = subprocess.run(execute_on_head)
```

Comment:
as Kai-Hsun mentioned, there are a bunch of gotchas with string escaping etc. here

there is existing code for `ray job submit` that handles it -- let's reuse it